core/src/main/java/org/elasticsearch/index/query/ConstantScoreQueryBuilder.java

```diff
@@ -66,20 +70,78 @@ public ConstantScoreQueryBuilder boost(float boost) {
         return this;
     }
 
+    /**
+     * @return the boost factor
+     */
+    public float boost() {
+        return this.boost;
+    }
+
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
         builder.field("filter");
-        filterBuilder.toXContent(builder, params);
+        queryBuilder.toXContent(builder, params);
+        builder.field("boost", boost);
+        builder.endObject();
+    }
+
+    @Override
+    public Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+        Query innerFilter = queryBuilder.toQuery(parseContext);
+        if (innerFilter == null ) {
+            // no inner query, we ignore this and return no docs
+            return Queries.newMatchNoDocsQuery();
```

Comment:
I think there's two different things that we need to decide here, sorry about the confusion:

1) whether we want to rely on validate or not, and assume things are not null where they shouldn't. This is potentially dangerous, I think I agree that we should stay on the safe side for now although it might require double checks

2) whether we should enforce toQuery to never return null itself and rely on this documented behaviour. In that case we wouldn't need this null check here. But the case where no innerFilter is specified falls into the previous case, and would currently throw NPE at the first line of the method

> innerFilter can still be null if we haven't refactored it

sure but we test only what we refactored so that should not be a huge problem. To me either we should either say the result can be null and we just live with that or we document and guarantee null is not an option, the result always needs to be a proper lucene query. I think we shouldn't try and do both. If existing queries currently return null they won't anymore once we migrated them.